In [1]:
import h5py
import pandas as pd

C:\Users\bananth\AppData\Local\Continuum\anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [2]:
filename = 'SVHN_single_grey1.h5'

In [3]:
f = h5py.File(filename,'r')

In [4]:
print("Keys: %s" % f.keys())
a_group_key = list(f.keys())

Keys: KeysView(<HDF5 file "SVHN_single_grey1.h5" (mode r)>)


In [5]:
a_group_key

['X_test', 'X_train', 'X_val', 'y_test', 'y_train', 'y_val']

In [6]:
X_test = f['X_test']

In [7]:
X_train = f['X_train']

In [8]:
X_val = f['X_val']

In [9]:
y_test = f['y_test']

In [10]:
y_train = f['y_train']

In [11]:
y_val = f['y_val']

In [12]:
for key in f:
    d = f[key]
    print(d)
    print(d.shape)

<HDF5 dataset "X_test": shape (18000, 32, 32), type "<f4">
(18000, 32, 32)
<HDF5 dataset "X_train": shape (42000, 32, 32), type "<f4">
(42000, 32, 32)
<HDF5 dataset "X_val": shape (60000, 32, 32), type "<f4">
(60000, 32, 32)
<HDF5 dataset "y_test": shape (18000,), type "|u1">
(18000,)
<HDF5 dataset "y_train": shape (42000,), type "|u1">
(42000,)
<HDF5 dataset "y_val": shape (60000,), type "|u1">
(60000,)


In [13]:
y_train.value.max()

9

In [14]:
from sklearn.neighbors import KNeighborsClassifier


In [15]:
X_train.value[0]

array([[ 33.0704,  30.2601,  26.852 , ...,  71.4471,  58.2204,  42.9939],
       [ 25.2283,  25.5533,  29.9765, ..., 113.0209, 103.3639,  84.2949],
       [ 26.2775,  22.6137,  40.4763, ..., 113.3028, 121.775 , 115.4228],
       ...,
       [ 28.5502,  36.212 ,  45.0801, ...,  24.1359,  25.0927,  26.0603],
       [ 38.4352,  26.4733,  23.2717, ...,  28.1094,  29.4683,  30.0661],
       [ 50.2984,  26.0773,  24.0389, ...,  49.6682,  50.853 ,  53.0377]],
      dtype=float32)

In [16]:
print(X_train.shape)
print(X_test.shape)
print(X_val.shape)
print(y_train.shape)
print(y_test.shape)

(42000, 32, 32)
(18000, 32, 32)
(60000, 32, 32)
(42000,)
(18000,)


In [17]:
y_train.shape

(42000,)

In [18]:
nsamples, nx, ny = X_train.shape
n1samples,n1x,n1y = X_test.shape
n2samples,n2x,n2y = X_val.shape

In [19]:
import numpy as np

In [20]:
X_train_new = X_train.value.reshape((nsamples,nx*ny))
X_test_new = X_test.value.reshape((n1samples,n1x*n1y))
X_val_new = X_val.value.reshape((n2samples,n2x*n2y))

In [21]:
print(X_train_new.shape)
print(X_test_new.shape)
print(X_val_new.shape)
print(y_train.shape)
print(y_test.shape)

(42000, 1024)
(18000, 1024)
(60000, 1024)
(42000,)
(18000,)


In [22]:
knn_5 = KNeighborsClassifier(n_neighbors=5)

In [23]:
knn_5.fit(X_train_new,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [ ]:
y_pred_class_knn_5 = knn_5.predict(X_test_new)

In [ ]:
from sklearn import metrics
print("KNN ACCURACY for 5 Neighbors:",metrics.accuracy_score(y_test,y_pred_class_knn_5))
print('KNN CLASSIFIER CONFUSTION MATRIX - 5 Neighbors:',pd.crosstab(y_test,y_pred_class_knn_5))

In [68]:
from sklearn.metrics import accuracy_score, classification_report, f1_score, confusion_matrix

In [69]:
print("Accuracy score %5.2f " %(accuracy_score(y_true=y_test, y_pred=y_pred_class_knn_5)))
print("F1 score is %5.2f" %(f1_score(y_true=y_test, y_pred=y_pred_class_knn_5, average='weighted')))
print("Classification report")
print(classification_report(y_test, y_pred_class_knn_5))
print("Confusion matrix ")
print(confusion_matrix(y_test, y_pred_class_knn_5))

Accuracy score  0.49 
F1 score is  0.49
Classification report
             precision    recall  f1-score   support

          0       0.38      0.66      0.48      1814
          1       0.44      0.70      0.54      1828
          2       0.55      0.52      0.53      1803
          3       0.42      0.39      0.41      1719
          4       0.67      0.61      0.64      1812
          5       0.48      0.36      0.41      1768
          6       0.47      0.38      0.42      1832
          7       0.72      0.58      0.64      1808
          8       0.42      0.33      0.37      1812
          9       0.52      0.36      0.43      1804

avg / total       0.51      0.49      0.49     18000

Confusion matrix 
[[1201   60   44   37   45   45  129   31  109  113]
 [ 125 1285   83   81   84   36   36   58   18   22]
 [ 126  252  932  115   44   44   28  128   69   65]
 [ 164  255  166  677   54  157   36   46   92   72]
 [ 148  256   47   61 1101   30   64   17   48   40]
 [ 196  179   91

In [22]:
import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [23]:
#Let us normalize the data
X_train_new = X_train_new/1024
X_test_new = X_test_new/1024
X_val_new = X_val_new/1024

In [24]:
print(X_train_new.shape)
print(X_test_new.shape)
print(X_val_new.shape)

(42000, 1024)
(18000, 1024)
(60000, 1024)


In [25]:
#Let us convert Y values to categorical
y_train_new = tf.keras.utils.to_categorical(y_train, num_classes=10)
y_test = tf.keras.utils.to_categorical(y_test, num_classes=10)
y_val = tf.keras.utils.to_categorical(y_val, num_classes=10)

In [26]:
y_train_new.shape

(42000, 10)

In [27]:
y_train.shape

(42000,)

In [45]:
#Let us create a model
model = tf.keras.models.Sequential()

In [46]:
#Add batch normalization
#model.add(tf.keras.layers.BatchNormalization())
#Add a dense layer with 200 neurons
tf.keras.layers.Dense(units=200, activation='relu', input_shape=(1024,))

In [47]:
#Let us add few more layers with units=100, 50, 20 and output layer with unit as 10
model.add(tf.keras.layers.Dense(units=150, activation='relu'))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Dense(units=70, activation='relu'))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Dense(units=30, activation='relu'))
model.add(tf.keras.layers.BatchNormalization())
model.add(tf.keras.layers.Dense(units=10, activation='softmax'))

In [50]:
#Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy',precision,recall])

In [51]:
#Fit the data and execute the model
model.fit(x=X_train_new, y=y_train_new, batch_size=64, epochs=30, validation_data=(X_val_new, y_val))

Train on 42000 samples, validate on 60000 samples
Epoch 1/30
42000/42000 [==============================] - 10s 230us/step - loss: 2.3039 - acc: 0.1005 - precision: 0.0000e+00 - recall: 0.0000e+00 - val_loss: 2.3028 - val_acc: 0.1000 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
Epoch 2/30
42000/42000 [==============================] - 7s 177us/step - loss: 2.2727 - acc: 0.1240 - precision: 0.1676 - recall: 0.0044 - val_loss: 2.1571 - val_acc: 0.1761 - val_precision: 0.1013 - val_recall: 0.0162
Epoch 3/30
42000/42000 [==============================] - 7s 176us/step - loss: 2.0692 - acc: 0.2117 - precision: 0.6555 - recall: 0.0323 - val_loss: 1.9970 - val_acc: 0.2386 - val_precision: 0.1231 - val_recall: 0.0338
Epoch 4/30
42000/42000 [==============================] - 7s 173us/step - loss: 1.9636 - acc: 0.2499 - precision: 0.6860 - recall: 0.0427 - val_loss: 1.9454 - val_acc: 0.2517 - val_precision: 0.2038 - val_recall: 0.0412
Epoch 5/30
42000/42000 [=============================

In [52]:
y_pred = model.predict(X_test_new)

In [53]:
model.evaluate(X_test_new, y_test)

18000/18000 [==============================] - 1s 77us/step


[1.21189995320638, 0.6026111111111111, 0.7751187131669787, 0.45422222222222225]

### Observations:
    KNN was much slower for k value =5.
    Neural networks, without much of tuning still managed to get good accuracy along with high accuracy
    Printed Recall & Precision also while running the model. with 75% and 45% Recall for the Testing Accuracy as well.